common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/AbstractGatewaySessionHandler.java

```diff
@@ -787,15 +790,37 @@ public void onSuccess(Void dummy) {
                 log.trace("[{}][{}][{}][{}] Published msg: [{}]", gateway.getTenantId(), gateway.getDeviceId(), sessionId, deviceName, msg);
                 if (msgId > 0) {
                     ctx.writeAndFlush(MqttTransportHandler.createMqttPubAckMsg(deviceSessionCtx, msgId, ReturnCode.SUCCESS));
+                } else {
+                    log.trace("[{}][{}][{}] Wrong msg id: [{}]", gateway.getTenantId(), gateway.getDeviceId(), sessionId, msg);
+                    ctx.writeAndFlush(MqttTransportHandler.createMqttPubAckMsg(deviceSessionCtx, msgId, ReturnCode.UNSPECIFIED_ERROR));
+                    closeDeviceSession(deviceName, ReturnCode.MALFORMED_PACKET);
                 }
             }
 
             @Override
             public void onError(Throwable e) {
                 log.trace("[{}][{}][{}] Failed to publish msg: [{}] for device: [{}]", gateway.getTenantId(), gateway.getDeviceId(), sessionId, msg, deviceName, e);
+                if (e instanceof TbRateLimitsException) {
+                    closeDeviceSession(deviceName, ReturnCode.MESSAGE_RATE_TOO_HIGH);
+                } else {
+                    closeDeviceSession(deviceName, ReturnCode.UNSPECIFIED_ERROR);
+                }
                 ctx.close();
             }
         };
     }
 
+
+    private void closeDeviceSession(String deviceName, ReturnCode returnCode) {
+        try {
+            if (MqttVersion.MQTT_5.equals(deviceSessionCtx.getMqttVersion())) {
+                MqttTransportAdaptor adaptor = deviceSessionCtx.getAdaptor(TopicType.V1);
```

Comment:
Could you please describe why we always use `deviceSessionCtx.getAdaptor(TopicType.V1);` instead of `deviceSessionCtx.getPayloadAdaptor()`